### PR TITLE
feat: validate @range(min, max) for Int type

### PR DESCRIPTION
**Pull request by mrl5**

### CODE REVIEW

**Comment by mrl5** [comment]
I tried to `vimdiff` my `index.exo_ir`. Comparing output from upstream vs output from my changes

I don't see the `range` there so I'm not sure if I'm missing something or I should validate it differently

**Comment by ramnivas** [comment]
Not sure about the vimdiff way, but may be you can printing at, say, https://github.com/exograph/exograph/blob/main/crates/postgres-subsystem/postgres-model/src/types.rs#L282. This is where you would be modifying next, anyway.

**Comment by mrl5** [comment]
hmm indeed it's there when I `println!("{:?}", &self.type_props)`. Let's see how we can bridge that with runtime in following PRs...

File: integration-tests/field-validation/tests/create-example-with-out-of-range-value.exotest [hunk 1]
```diff
@@ -0,0 +1,21 @@
+operation: |
```
Comment (ramnivas) [inline integration-tests/field-validation/tests/create-example-with-out-of-range-value.exotest:1]: Please add a few positive tests, especially on the edges of the range.

File: crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs [hunk 2]
```diff
@@ -411,3 +423,63 @@ impl<'a> ArgumentValidator<'a> {
...
 }
+
+fn get_schema_directives(schema: &Schema, base_type_name: &str) -> Vec<ConstDirective> {
+    schema
+        .type_definitions
+        .iter()
+        .filter(|td| matches!(&td.kind, TypeKind::Object(_)))
+        .flat_map(|td| {
+            td.fields()
+                .unwrap()
```
Comment (mrl5) [inline crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs:434]: most likely also this should be improved

Comment (ramnivas) [reply crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs:434]: Yes, you can make this return `Result<Vec<ConstDirective>, ValidationError>`. You may need to introduce a variation in `ValidationError` (something like `InvalidValidationInfo`).

File: crates/core-subsystem/core-resolver/src/validation/validation_error.rs [hunk 3]
```diff
@@ -74,6 +74,14 @@ pub enum ValidationError {
 
     #[error("Selection set too deep")]
     SelectionSetTooDeep(Pos),
+
+    #[error("Invalid value for '{value_name}': {range_detail}, {value_detail}")]
+    ValueOutOfRange {
+        value_name: String,
+        range_detail: String,
```
Comment (ramnivas) [inline crates/core-subsystem/core-resolver/src/validation/validation_error.rs:81]: May be not keep these as stings (instead keep the raw values) and let the formatting in the `error` macro take care.

File: integration-tests/field-validation/tests/create-example-with-out-of-range-value.exotest (hunk 1 above)
Comment (ramnivas) [reply integration-tests/field-validation/tests/create-example-with-out-of-range-value.exotest:1]: Let's also add tests for other mutations

File: crates/postgres-subsystem/postgres-model/src/types.rs [hunk 4]
```diff
@@ -177,6 +177,12 @@ pub struct PostgresField<CT> {
     pub dynamic_default_value: Option<ContextSelection>,
     pub readonly: bool,
     pub access: Access,
+    pub type_props: Option<TypeProps>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum TypeProps {
```
Comment (ramnivas) [inline crates/postgres-subsystem/postgres-model/src/types.rs:184]: Let's rename this to something like: `TypeValidation` or `TypeValidationInfo` or `ValidationInfo`

File: crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs [hunk 5]
```diff
@@ -411,3 +423,63 @@ impl<'a> ArgumentValidator<'a> {
...
+                value_name: "range".into(),
+                range_detail: format!("min = {}", r),
+                value_detail: format!("provided value = {}", value),
+                pos,
+            });
+        }
+    }
+
+    if let Some(r) = int_max {
+        if value > r {
```
Comment (ramnivas) [inline crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs:474]: How do we want to treat `max` (inclusive?).

File: crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs [hunk 6]
```diff
@@ -423,13 +423,14 @@ impl<'a> ArgumentValidator<'a> {
...
 fn get_schema_directives(schema: &Schema, base_type_name: &str) -> Vec<ConstDirective> {
+    let placeholder: Vec<Positioned<FieldDefinition>> = vec![];
     schema
         .type_definitions
         .iter()
         .filter(|td| matches!(&td.kind, TypeKind::Object(_)))
         .flat_map(|td| {
             td.fields()
-                .unwrap()
+                .unwrap_or(&placeholder)
```
Comment (ramnivas) [inline crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs:433]: nit: you can use https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_default here to simplify it further

Comment (mrl5) [reply crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs:433]: I think I can't unfortunately:
```
the trait bound `&Vec<Positioned<async_graphql_parser::types::FieldDefinition>>: Default` is not satisfied: the trait `Default` is not implemented for `&Vec<Positioned<async_graphql_parser::types::FieldDefinition>>` 
```

Comment (ramnivas) [reply crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs:433]: I see!

File: integration-tests/field-validation/tests/create-example-with-out-of-range-value.exotest [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+              "locations": [
+                {
+                  "line": 2,
+                  "column": 23
+                }
+              ]
+            }
+          ]
+        }
+  - operation: |
```
Comment (mrl5) [inline integration-tests/field-validation/tests/create-example-with-out-of-range-value.exotest:44]: I forgot to remove it, it's already tested in `create-example.exotest`

File: crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs [hunk 8]
```diff
@@ -432,7 +437,10 @@ fn get_schema_directives(schema: &Schema, base_type_name: &str) -> Vec<ConstDire
             td.fields()
                 .unwrap_or(&placeholder)
                 .iter()
-                .filter(|x| x.node.ty.node.base.to_string() == base_type_name)
+                .filter(|x| {
+                    x.node.name.node.as_str() == argument_definition
+                        && x.node.ty.node.base.to_string() == base_type_name
```
Comment (mrl5) [inline crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs:442]: most likely names are unique and proably we don't need this check, no?

File: integration-tests/field-validation/tests/update-example.exotest [hunk 9]
```diff
@@ -0,0 +1,33 @@
+stages:
+  - operation: |
+        mutation {
+          updateExample(id: 1, data: { limit: 0 }) {
```
Comment (ramnivas) [inline integration-tests/field-validation/tests/update-example.exotest:4]: Adding `age` here with values outside the `limit` range will make these tests even better

File: crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs [hunk 10]
```diff
@@ -432,7 +432,7 @@ fn get_schema_directives(schema: &Schema, base_type_name: &str) -> Vec<ConstDire
             td.fields()
                 .unwrap_or(&placeholder)
                 .iter()
-                .filter(|x| x.node.ty.node.base.to_string() == base_type_name)
+                .find(|x| x.node.name.node.as_str() == argument_name)
```
Comment (ramnivas) [inline crates/core-subsystem/core-resolver/src/validation/arguments_validator.rs:435]: Perfect! Indeed, a field name is unique within a type (we check for it and issue an error if that is not the case).

**Pull request merged**